### ISSUE TITLE
Add token consumption and multi-viewport calibration signals

### ISSUE THREAD

**Issue by let-sunny** [let-sunny/canicode#108]
## Problem

The only calibration signal is visual-compare (static screenshot pixel diff). This works for structure/token rules but systematically undervalues component and behavior rules:

- **Component rules**: visual output is identical whether componentized or not → always appears "overscored"
- **Behavior rules**: static screenshots don't resize/scroll → overflow/truncation/responsive rules can't be validated

From calibration evidence: `detached-instance` 7× overscored / 0× validated, `text-truncation-unhandled` 5× overscored / 0× validated.

Equal category weights (1.0 each) are correct for the project's maintainability goal, but the measurement can't validate this — the problem is measurement, not weighting (see closed #107).

## Methodology: OAT Sensitivity Analysis + Extended Thinking

Our approach is based on [OAT (One-at-a-Time) Sensitivity Analysis](https://en.wikipedia.org/wiki/Sensitivity_analysis) — a standard method from engineering/statistics for measuring individual factor importance by changing one input at a time while holding others at baseline.

Applied to rule validation:

```
Baseline:  full design tree → AI implements (extended thinking) → thinking T₀, visual V₀
Ablation:  strip rule-related properties → AI implements → thinking T₁, visual V₁

ΔT = T₁ - T₀  (thinking increase = how much harder it got)
ΔV = V₀ - V₁  (visual decrease = how much worse the result got)
```

Why ablation delta over absolute measurement:
- **No circular reasoning** — strips design data, not rule logic
- **Variance cancelled out** — relative measurement between paired runs
- **Direct attribution** — delta is caused by the specific stripped property
- **All categories testable** — every rule has strippable properties

Related methods for future consideration:
- [Morris Method](https://en.wikipedia.org/wiki/Morris_method) — OAT upgrade that detects interaction effects
- [Shapley Values](https://christophm.github.io/interpretable-ml-book/shapley.html) — game-theory fair attribution (feasible at category level: 2^5 = 32 combinations)
- [Permutation Feature Importance](https://scikit-learn.org/stable/modules/permutation_importance.html) — shuffles instead of removes

### Step 0: Pilot validation (~$40)

**Goal**: Validate that ΔT (thinking token delta) is a stable, usable signal before investing in full OAT.

**3 rules × 2 fixtures × 2 repetitions = 16 runs**

| Rule | Category | Expected ΔT | Expected ΔV | Role |
|---|---|---|---|---|
| `no-auto-layout` | structure | large | large | positive control |
| `missing-component` | component | large | small | the blind spot we're solving |
| `default-name` | naming | ≈ 0 | ≈ 0 | negative control |

Pass criteria:
- ΔT coefficient of variation < 30% between repeated runs
- ΔV direction matches expectation for all 3 rules
- No strip errors or design tree corruption

**Contingency based on pilot results:**

```
ΔT stable?  ──Yes──→ ΔT + ΔV both used (best case) → Step 1
    │
    No
    │
ΔV alone valid? ──Yes──→ ΔV-only OAT + component rules handled separately
    │
    No
    │
Strip issue? ──Yes──→ Switch from removal to perturbation (inject wrong values)
    │
    No
    │
    └──→ Abandon OAT, fall back to manual evaluation (Alternative 2)
```

### Step 1: Full OAT (29 rules × N fixtures, ~$225)

Run ablation for all 29 rules. For each rule:
1. Strip rule-specific properties from design tree
2. Run Converter with extended thinking
3. Record ΔT, ΔV
4. Rank rules by empirical importance

Output: empirical importance ranking compared against current rule scores.

### Step 2: Interaction check (data-driven)

Not pre-defined — driven by Step 1 results:
- Rules where ΔV is **unexpectedly small** despite high expected importance
- Pattern: individually weak delta + high expected importance = interaction candidate
- Test suspicious pairs with simultaneous removal
- If interactions are significant → consider upgrading to Morris method

### Step 3: Category-level Shapley (2^5 = 32 combinations, ~$250)

5 categories → 32 combinations → feasible and directly actionable:
- Fair attribution of each category's contribution to implementation difficulty
- Directly informs `CATEGORY_WEIGHT` in `scoring.ts` (currently all 1.0)
- Replaces assumption with empirical data

## What to strip per rule

**Structure**
| Rule | Properties to strip |
|---|---|
| `no-auto-layout` | `layoutMode`, flex-direction, gap, align-items, justify-content |
| `absolute-position-in-auto-layout` | `layoutPositioning` on children |
| `fixed-size-in-auto-layout` | `layoutSizingHorizontal`, `layoutSizingVertical` |
| `missing-size-constraint` | `constraints`, min/max width/height |
| `missing-responsive-behavior` | `layoutGrow`, `layoutAlign` |
| `group-usage` | Convert GROUPs to FRAMEs (or vice versa) |
| `deep-nesting` | Flatten intermediate wrapper nodes |
| `z-index-dependent-layout` | Remove overlap info |
| `unnecessary-node` | Remove single-child passthrough nodes |

**Token**
| Rule | Properties to strip |
|---|---|
| `raw-color` | Remove style references, keep raw hex values |
| `raw-font` | Remove font style references |
| `inconsistent-spacing` | Randomize `itemSpacing` values |
| `magic-number-spacing` | Replace grid-aligned spacing with arbitrary values |
| `raw-shadow` | Remove shadow style references |
| `raw-opacity` | Remove opacity style references |
| `multiple-fill-colors` | Add extra fill colors |

**Component**
| Rule | Properties to strip |
|---|---|
| `missing-component` | Remove `[component: ...]` annotations, inline component trees |
| `detached-instance` | Remove `componentId` from instances |
| `missing-component-description` | Remove component description metadata |
| `variant-structure-mismatch` | Alter variant child structure |

**Behavior**
| Rule | Properties to strip |
|---|---|
| `text-truncation-unhandled` | Remove `textAutoResize`, `textTruncation` |
| `overflow-behavior-unknown` | Remove `clipsContent`, `overflowDirection` |
| `wrap-behavior-unknown` | Remove `layoutWrap` |

**Naming**
| Rule | Properties to strip |
|---|---|
| `default-name` | Replace names with "Frame 1", "Group 2" |
| `non-semantic-name` | Replace names with "element", "container" |
| `inconsistent-naming-convention` | Mix camelCase/kebab-case randomly |
| `numeric-suffix-name` | Append " 2", " 3" to names |
| `too-long-name` | Extend names to 60+ characters |

## Out of Scope

- **Naming rules**: no special treatment needed. OAT will empirically validate whether naming information matters for AI implementation (we expect ΔV ≈ 0, confirming current low scores).

## Related

- #104 (calculatedScore in scoring) — merged
- #105 (density + depth weighting) — closed
- #106 (diversity severity weighting) — merged as #111
- #107 (category weights — correct, measurement is the problem) — closed

### Core methodology

- **Ablation Study (AI)**: Newell, A. (1975). *A Tutorial on Speech Understanding Systems*. In Speech Recognition: Invited Papers of the IEEE Symposium. — [Wikipedia: Ablation (artificial intelligence)](https://en.wikipedia.org/wiki/Ablation_(artificial_intelligence))
  - Originated from experimental neuropsychology (brain lesion studies). First applied to AI by Allen Newell in 1975 to analyze speech recognition systems.
  - Modern treatment: Meyes, R. et al. (2019). *Ablation Studies in Artificial Neural Networks*. — [arXiv:1901.08644](https://arxiv.org/abs/1901.08644)

- **OAT (One-at-a-Time) Sensitivity Analysis**: Daniel, C. (1973). *One-at-a-Time Plans*. Journal of the American Statistical Association, 68(342), 353–360. — [DOI:10.1080/01621459.1973.10482433](https://www.tandfonline.com/doi/abs/10.1080/01621459.1973.10482433)
  - Standard method for screening factor importance. Change one input at a time, measure output delta. Known limitation: cannot detect interaction effects between factors.

### Interaction detection (Step 2 upgrade candidate)

- **Morris Method**: Morris, M. D. (1991). *Factorial Sampling Plans for Preliminary Computational Experiments*. Technometrics, 33(2), 161–174. — [DOI:10.1080/00401706.1991.10484804](https://www.tandfonline.com/doi/abs/10.1080/00401706.1991.10484804)
  - Extends OAT with multiple randomized trajectories. Detects interaction effects while remaining linear in the number of factors. Suitable if Step 2 reveals significant interactions.

### Fair attribution (Step 3)

- **Shapley Values**: Shapley, L. S. (1953). *A Value for n-Person Games*. In Contributions to the Theory of Games II, Princeton University Press, 307–317. — [RAND P-295](https://www.rand.org/pubs/papers/P295.html)
  - Game-theory foundation for fair contribution attribution. Only method satisfying all four fairness axioms (efficiency, symmetry, additivity, dummy).

- **SHAP**: Lundberg, S. M. & Lee, S.-I. (2017). *A Unified Approach to Interpreting Model Predictions*. NeurIPS 2017. — [arXiv:1705.07874](https://arxiv.org/abs/1705.07874)
  - Practical implementation of Shapley values for ML model interpretation. Unifies 6 existing methods. Applicable at category level (2^5 = 32 combinations).

### Feature importance baseline

- **Permutation Feature Importance**: Breiman, L. (2001). *Random Forests*. Machine Learning, 45(1), 5–32. — [PDF](https://www.stat.berkeley.edu/~breiman/randomforest2001.pdf)
  - Measures importance by shuffling feature values and observing performance degradation. Our approach uses removal (ablation) instead of shuffling (permutation), which is a stronger intervention.
  - Model-agnostic extension: Fisher, A., Rudin, C. & Dominici, F. (2019). *All Models are Wrong, but Many are Useful*. JMLR. — [arXiv:1801.01489](https://arxiv.org/abs/1801.01489)
